### ISSUE TITLE
[Suggestion] JavaScript Script support (or make meteor addons more flexible)

### ISSUE THREAD

**Issue by Cypphi** [MeteorDevelopment/meteor-client#467]
Add JavaScript script support so we can code our own modules while not having to reload all the time. It would be very useful and way better than the addons since the addons are separate mods we have to reload all the time.

If the option above isn't gonna be implemented, please allow meteor addons to modify anything of meteor itself. By this I'm talking about themes, extra module options and what not.....

**Comment by MineGame159** [comment]
Addons can already add themes, and for modifying existing modules thats extremely tricky to do, not even possible without changing how every module fundamentally works

**Comment by seasnail8169** [comment]
fyi you can overwrite a module by naming it the same thing and putting it in the same category

**Comment by MineGame159** [comment]
Well yeah but it has some problems

**Comment by CountBleck** [comment]
This is just my own opinion, but I would like a JS API, since then I don't need to learn Java, compile anything, and make a bunch of classes and files, when I only want to make a simple script that I only need once.

**Comment by BobIsMyManager** [comment]
Also if js you can add stuff and maybe do .reload so you wouldn’t have to restart your game

**Comment by Declipsonator** [comment]
> This is just my own opinion, but I would like a JS API, since then I don't need to learn Java, compile anything, and make a bunch of classes and files, when I only want to make a simple script that I only need once.

I think some other things should be added and some bugs be fixed before anyone thinks abt adding a javascript api

**Comment by Cypphi** [comment]
> > This is just my own opinion, but I would like a JS API, since then I don't need to learn Java, compile anything, and make a bunch of classes and files, when I only want to make a simple script that I only need once.
> 
> I think some other things should be added and some bugs be fixed before anyone thinks abt adding a javascript api

Yeah very true, or just make addons more flexible instead of adding a whole new scripting system

**Comment by BobIsMyManager** [comment]
Scripts are much easier to code tho

**Comment by CountBleck** [comment]
Yeah, scripts are easier to make without compiling and stuff. However, you would somehow have to add a JavaScript engine or some other interpreter (depends on what language is used to write scripts in).
In the case of JavaScript, you would need to embed V8, Spidermonkey, or ChakraCore somehow. And none of those are written in Java; they use C++. So probably no scripting APIs using JS any time soon.

**Comment by link-discord** [comment]
compile it with the typescript compiler to es5 then lol

**Comment by JFronny** [comment]
Technically, theres GraalJS, but I agree that this doesn't seem worthwhile, especially when there are fully-featured fabric mods solely focused on this

**Comment by CountBleck** [comment]
> compile it with the typescript compiler to es5 then lol

You mean Babel. Also, that's inconvenient for making scripts when you need them. Moreover, that adds a dependency on Node.js, which is stupid for a Minecraft mod.

**Comment by MukjepScarlet** [comment]
> Technically, theres GraalJS, but I agree that this doesn't seem worthwhile, especially when there are fully-featured fabric mods solely focused on this

LiquidBounce has GraalJS now. But it makes the size of jar file doubled or more. And because LiquidBounce has many usages of Kotlin language features, the script writer should read the utils written in Kotlin and try to understand its ASM(Java) interop with JS. To be honest I rarely see anyone use it. I'd rather write the feature I want with Kotlin.

**Comment by MukjepScarlet** [comment]
I think Groovy or Kotlin script will be much better here. But I never know how they work.

**Comment by Cypphi** [comment]
> I think Groovy or Kotlin script will be much better here. But I never know how they work.

Possibly, but judging by how long this has been open. It doesn't have priority. And honestly back then I didn't even know what I was talking about, still sometimes don't lol

**Comment by MukjepScarlet** [comment]
lol I think it's even easier to write a mod which depends on the client

**Comment by Cypphi** [comment]
> lol I think it's even easier to write a mod which depends on the client

It is undoubtedly easier. The addon system is nice the way it is tbh. Nowadays I don't even use addons anymore and just use my own base
